src/Memory/MemoryUtilsTest.cpp

```diff
@@ -0,0 +1,110 @@
+// Copyright (c) 2021 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include <absl/strings/substitute.h>
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+
+#include <string>
+
+#include "MemoryUtils.h"
+
+using orbit_grpc_protos::TotalMemoryUsageInfo;
+
+namespace orbit_memory {
+
+TEST(MemoryUtils, ParseMemInfo) {
+  const int kMemTotal = 16396576;
+  const int kMemFree = 11493816;
+  const int kMemAvailable = 14378752;
+  const int kBuffers = 71540;
+  const int kCached = 3042860;
+  const int kMissingInfo = -1;
+
+  const std::string kValidMeminfo =
+      absl::Substitute(R"(MemTotal:       $0 kB
+MemFree:        $1 kB
+MemAvailable:   $2 kB
+Buffers:        $3 kB
+Cached:         $4 kB
+SwapCached:            0 kB
+Active:          3350508 kB
+Inactive:        1190988 kB
+Active(anon):    1444908 kB
+Inactive(anon):      516 kB
+Active(file):    1905600 kB
+Inactive(file):  1190472 kB
+Unevictable:       56432 kB
+Mlocked:           56432 kB
+SwapTotal:       1953788 kB
+SwapFree:        1953788 kB
+Dirty:               492 kB
+Writeback:             0 kB
+AnonPages:       1326896 kB
+Mapped:           716656 kB
+Shmem:               796 kB
+KReclaimable:      84864 kB
+Slab:             194376 kB
+SReclaimable:      84864 kB
+SUnreclaim:       109512 kB
+KernelStack:       24724 kB
+PageTables:        13164 kB
+NFS_Unstable:          0 kB
+Bounce:                0 kB
+WritebackTmp:          0 kB
+CommitLimit:    10152076 kB
+Committed_AS:    6324736 kB
+VmallocTotal:   34359738367 kB
+VmallocUsed:       38264 kB
+VmallocChunk:          0 kB
+Percpu:             3072 kB
+HardwareCorrupted:     0 kB
+AnonHugePages:    782336 kB
+ShmemHugePages:        0 kB
+ShmemPmdMapped:        0 kB
+FileHugePages:         0 kB
+FilePmdMapped:         0 kB
+HugePages_Total:       0
+HugePages_Free:        0
+HugePages_Rsvd:        0
+HugePages_Surp:        0
+Hugepagesize:       2048 kB
+Hugetlb:               0 kB
+DirectMap4k:      201960 kB
+DirectMap2M:     5040128 kB
+DirectMap1G:    13631488 kB)",
+                       kMemTotal, kMemFree, kMemAvailable, kBuffers, kCached);
+
+  const std::string kPartialMeminfo = absl::Substitute(R"(MemTotal:       $0 kB
+MemFree:        $1 kB
+SwapCached:      0 kB)",
+                                                       kMemTotal, kMemFree);
+
+  const std::string kEmptyMeminfo = "";
+
+  std::optional<TotalMemoryUsageInfo> parsing_result;
+
+  // TODO(vickyliu): Check whether we are able to user testing::EqualsProto from the #include
```

Comment:
"to use"

Also you can have a Look at the "VulkanLayerControllerTest" line 132 to see gmock matchers in action.

Reply:
Thanks so much for providing the information about `Matcher`.

But I had some trouble when applying the `Field` method to the message. `VkExtensionProperties` is a struct so we can handle the filed in the way of `Field("specVersion", &VkExtensionProperties::specVersion, expected.specVersion)`. But I cannot process the filed of `SystemMemoryUsage` in a similar way, e.g.,  `Field("total_kb", &SystemMemoryUsage::total_kb(), expected.total_kb())`. There is an error about the second param, which says that I cannot access a non-static member without an object.

I removed the comments and keep the rest part as it is for now. I will try to address this problem later.

Reply:
Have you checked the gMock cookbook? https://github.com/google/googletest/blob/master/docs/gmock_cook_book.md#validating-a-member-of-an-object

But in general it is totally fine to leave it as it is. I'd don't think its worth to invest a lot of time here, other then learning the gMock API. Also, you could simply move the `EXPECT`s into a helper method.

Reply:
Thanks for providing this helpful resource for gMock. Also I moved the `EXPECT`s into a helper method called `CheckMemInfoParsingResult(...)`.